DataDog/dd-trace-js
# [serverless] Add S3 Span Pointers

### What does this PR do?
Adds span pointers in S3 for `putObject`, `copyObject`, and `completeMultipartUpload` requests.

Span pointers are similar to [Span Links](https://docs.datadoghq.com/tracing/trace_collection/span_links/), but for cases when it is impossible to pass the Trace ID and Span ID between the spans that need to be linked.

When the calculated hashes for the upstream and downstream lambdas match, the Datadog backend will automatically link the two traces together.

<img width="1338" alt="Screenshot 2024-11-11 at 1 29 11 PM" src="https://github.com/user-attachments/assets/10bab09f-c163-4281-b951-ca1c0cfbc18d">
When clicking on the linked span, a new tab opens linking to the downstream Lambda function that was triggered by this S3 object update.

### Motivation
This feature already exists in Python, and I'm working on adding it to all other runtimes (Node, .NET, Java, Golang).

### Testing Guidelines

Easy: Checkout [this span](https://ddserverless.datadoghq.com/functions?cloud=aws&config_serverless-auto-span-links=true&entity_view=lambda_functions&fromUser=false&graphType=flamegraph&panel_end=1731688980000&panel_paused=false&panel_start=1731688080000&shouldShowLegend=true&sort=time&sp=%5B%7B%22p%22%3A%7B%22entityId%22%3A%22aws-lambda-functions%2Bnhulston-javascript-upstream-dev-hello%2Bus-east-1%2B425362996713%22%7D%2C%22i%22%3A%22lambda-panel%22%7D%2C%7B%22p%22%3A%7B%22traceID%22%3A%226833201988596926111%22%2C%22selectedSpanID%22%3A%226833201988596926111%22%7D%2C%22i%22%3A%22trace-panel%22%7D%5D&spanID=6833201988596926111&spanViewType=span-links&text_search=nhulston&traceID=6833201988596926111&start=1731685380000&end=1731688980000&paused=false), enable the feature flag, and you will see that it's pointing to the downstream Lambda.

Run [this Lambda function](https://us-east-1.console.aws.amazon.com/lambda/home?region=us-east-1#/functions/nhulston-javascript-upstream-dev-hello?subtab=envVars&tab=testing) with the event payload
```json
{
  "shouldPutObject": false,
  "shouldCopyObject": false,
  "shouldMultipartUpload": false,
  "useV2": true
}
```
and change one of the bools to true. Also, try both with AWS SDK v2 and v3. Enable the span pointers feature flag, and check Datadog to ensure that the spans are properly linked.

I also added tests:
```shell
# Unit tests
yarn test packages/dd-trace/test/span_pointers.spec.js

# Integration tests
docker run -d -p 4566:4566 localstack/localstack
PLUGINS=aws-sdk SERVICES=localstack DD_DATA_STREAMS_ENABLED=true yarn test:plugins:ci 
```

### Additional Notes
- This handles the 'upstream' case, for Lambdas that make S3 requests. I have another PR in datadog-lambda-js to handle the 'downstream' case, for Lambdas that are triggered by S3 updates. https://github.com/DataDog/datadog-lambda-js/pull/587
- S3 and Dynamo are supported in Python. I will soon open another PR to add support for Dynamo.
- `packages/dd-trace/src/span_pointers.js` is a public API, since `datadog-lambda-js` will reuse the exported fields.
